lib/blockchain.py

```diff
@@ -419,89 +418,153 @@ def get_hash(self, height):
+            else:
+                new_target = self.bits_to_target(last.get('bits'))
+        elif NetworkConstants.REGTEST:
+            new_target = self.bits_to_target(last.get('bits'))
+        else:
+            first = self.read_header(height - difficulty_adjustment_interval())
+            target = self.bits_to_target(last.get('bits'))
 
-        actual_timespan = last.get('timestamp') - first.get('timestamp')
-        target_timespan = 14 * 24 * 60 * 60
-        actual_timespan = max(actual_timespan, target_timespan // 4)
-        actual_timespan = min(actual_timespan, target_timespan * 4)
+            actual_timespan = last.get('timestamp') - first.get('timestamp')
+            target_timespan = 14 * 24 * 60 * 60
+            actual_timespan = max(actual_timespan, target_timespan // 4)
+            actual_timespan = min(actual_timespan, target_timespan * 4)
 
-        new_target = min(NetworkConstants.POW_LIMIT_LEGACY, (target * actual_timespan) // target_timespan)
+            new_target = min(NetworkConstants.POW_LIMIT_LEGACY, (target * actual_timespan) // target_timespan)
 
         return new_target
 
-    def get_postfork_target(self, height, headers=None):
-        # Premine
-        if (height < NetworkConstants.FORK_HEIGHT + NetworkConstants.PREMINE_SIZE) or NetworkConstants.REGTEST:
+    def get_zawy_lwma_target(self, height, headers):
+        cur = headers[height] if height in headers else self.read_header(height)
+        last_height = (height - 1)
+        last = headers[last_height] if last_height in headers else self.read_header(last_height)
+
+        # Special testnet handling
+        if (NetworkConstants.TESTNET or NetworkConstants.REGTEST) and cur.get('timestamp') > last.get('timestamp') + NetworkConstants.POW_TARGET_SPACING * 2:
             new_target = NetworkConstants.POW_LIMIT
-            # Initial start (reduced difficulty)
-        elif height < NetworkConstants.FORK_HEIGHT + NetworkConstants.PREMINE_SIZE + NetworkConstants.POW_AVERAGING_WINDOW:
-            new_target = NetworkConstants.POW_LIMIT_START
-            # BTG default
+        elif NetworkConstants.REGTEST:
+            new_target = self.get_prev_target(height)
         else:
-            if headers is None:
-                headers = {}
+            total = 0
+            t = 0
+            j = 0
 
-            pow_limit = NetworkConstants.POW_LIMIT
-            height -= 1
-            last = headers[height] if height in headers else self.read_header(height)
+            # Loop through N most recent blocks.  "< height", not "<=".
+            # height-1 = most recently solved block
+            for i in range(height - NetworkConstants.ZAWY_AVERAGING_WINDOW, height):
+                cur = headers[i] if i in headers else self.read_header(i)
+                prev_height = (i - 1)
+                prev = headers[prev_height] if prev_height in headers else self.read_header(prev_height)
```

Comment:
There are lots of `headers[h] if h in headers else self.read_header(h)`. Maybe we should make it a common function.